fix: handling of default and star exports

mohd-akram opened the pull request:
Closes #68, Closes #77, Closes #62, Closes #60

mohd-akram commented:
I'm not sure, doesn't #86 ignore the `defaultAs` passed in? I don't think there's a reason to actually try and parse that and as I suspected it wasn't used.

timfish commented:
It can't rename `export default parseInt('1')` because it doesn't have a name. It's renamed further up the import grap. 

This PR may well fix it in another way but I'd try copying the tests across the confirm that.

mohd-akram commented:
Yes, clearly they do different things so either the test isn't covering it or something else. Not sure tbh. I added the tests here and they pass on my machine.

mohd-akram commented:
It turns out there are other problems with the handling of default exports. I reverted the default export changes from #53 as they stemmed from a misunderstanding that `export *` includes the default export as well. The tests were therefore wrong, so I removed the wrong tests. This likely fixes other cases as well.

timfish commented:
I've since opened #87, but it looks like even that can be trimmed down even more

timfish commented:
`setters` as an array won't suffice because when duplicate named exports are found, all of them need to be excluded from export.

mohd-akram commented:
It's a runtime error in Node.js if you do that:

```javascript
import { foo } from './entry.mjs'
         ^^^
SyntaxError: The requested module './entry.mjs' contains conflicting star exports for name 'foo'
```

where `entry.mjs` has `export * from './file1.js'` and `export * from './file2.js'` where `file1.js` == `file2.js` contents.

And if it's in the same file, it's an error as well:

```javascript
import { foo } from './file1.mjs'
import { foo } from './file2.mjs'

console.log(foo);
```

```javascript
import { foo } from './file2.mjs'
         ^^^

SyntaxError: Identifier 'foo' has already been declared
```

mohd-akram commented:
Star exports were also broken. All the tests currently pass on my machine.

timfish commented:
> It's a runtime error in Node.js if you do that:
> 
> ```js
> import { foo } from './entry.mjs'
>          ^^^
> SyntaxError: The requested module './entry.mjs' contains conflicting star exports for name 'foo'
> ```

Yes its a runtime error in Node, but it should also be a runtime error when using the loader hook. If you export a `foo`, there will no longer be a runtime error. That means if you find multiple named exports with the same name, they should all be excluded from the exports.

See my duplicate named exports tests here:
https://github.com/DataDog/import-in-the-middle/pull/87/files#diff-a39aaf48c07e70ea14099a7a5597088257a1b5ac55542df3a4d63e63e6546004

That's why I used this to remove duplicates:
https://github.com/DataDog/import-in-the-middle/pull/87/files#diff-e69a24a4c3746fa1ee96a78e12bb12d2dd4eb6e4cacbced2bf1f4084952681d9R119

mohd-akram commented:
Adding your tests to my PR results in a runtime error as expected:

```
not ok 7 test/hook/duplicate-exports.mjs
  ---
  stdout: ''
  stderr: >-
    file:///Users/mohamed/Downloads/code/import-in-the-middle/test/fixtures/duplicate.mjs?iitm=true:21
          let $foo = _.foo
              ^

    SyntaxError: Identifier '$foo' has already been declared
        at compileSourceTextModule (node:internal/modules/esm/utils:337:16)
        at ModuleLoader.moduleStrategy (node:internal/modules/esm/translators:163:18)
        at callTranslator (node:internal/modules/esm/loader:430:14)
        at ModuleLoader.moduleProvider (node:internal/modules/esm/loader:436:30)

    Node.js v22.2.0
```

timfish commented:
That's not the same runtime error, that's because the emitted iitm code contains duplicate identifiers.

You will also get that error if you do this which should not error:
`import * as lib from '/test/fixtures/duplicate.mjs'`

mohd-akram commented:
Okay, it seems in Node.js there is no runtime error if you do an `import *`, but if you `import foo` then it's an error. So yeah, this should be handled.

timfish commented:
Assuming all the tests pass on all node version this simplifies a lot more than my PR #87 so I'd favour this. 

As @jsumners-nr says [here](https://github.com/DataDog/import-in-the-middle/pull/87#discussion_r1618957420) though, this PR effectively reverts a PR that was supposed to fix something.

I think before we can even consider landing a PR like this with huge changes, we need to improve the tests with some real-world troublesome libraries.

timfish commented on hook.js:
```diff
@@ -192,69 +163,24 @@ async function processModule ({
-    .pop()
-    .replace(/(.+)\.(?:js|mjs)$/, '$1')
-    .replaceAll(/(-.)/g, x => x[1].toUpperCase())
+  return { setters }
```
Maybe just
```suggestion
  return setters
```

mohd-akram replied:
Thanks.

timfish commented on hook.js:
```diff
@@ -312,61 +238,25 @@ function createHook (meta) {
     if (hasIitm(url)) {
       const realUrl = deleteIitm(url)
-      const { imports, namespaces, setters: mapSetters } = await processModule({
+      const { setters } = await processModule({
```
```suggestion
      const setters = await processModule({
```

timfish commented:
Looks good!

Rather than me opening another PR to add more tests, do you think it's worth adding this patch to your PR?
[import-in-the-middle-0190d24-Add tests.patch](https://github.com/DataDog/import-in-the-middle/files/15488201/import-in-the-middle-0190d24-Add.tests.patch)

The `got` test passes on `main` and this PR fixes the other 3.

jsumners-nr commented:
https://github.com/newrelic/node-newrelic/issues/1920 is the issue where most of the discoveries come from. So it looks like the following should all work:

```js
import got from 'got'
typeof got.extend === 'function'
```

```js
import { defineScript } from 'redis'
typeof defineScript === 'function'
```

```js
import OpenAI from 'openai'
typeof OpenAI === 'function'
```

timfish commented:
@mohd-akram If you update you PR description to include the following, it'll auto-close them when this gets merged:

Closes https://github.com/DataDog/import-in-the-middle/issues/68, Closes https://github.com/DataDog/import-in-the-middle/issues/77, Closes https://github.com/DataDog/import-in-the-middle/issues/62, Closes https://github.com/DataDog/import-in-the-middle/issues/60...

mohd-akram commented:
@timfish Thanks, updated. There seems to be some problem with the `openai.mjs` test though, possibly an issue with imhotap. Running `imhotap --files test/hook/openai.mjs` in Node.js 16 even if it includes nothing but `import OpenAI from 'openai'` results in a `not ok` whereas `node test/hook/openai.mjs` exits with code 0 fine. Shall I remove it for now?

timfish commented:
What version of Node does the `openai` test pass on?

You can limit the test version by prepending a version the test file name:
`./test/hook/v18.19-openai.mjs`

mohd-akram commented:
Thanks, indeed it doesn't work before 18.19.

mohd-akram commented:
Had to move the other tests too. Looks good now.

AbhiPrasad commented on test/fixtures/default-class.mjs:
Why are we removing this and `test/fixtures/snake_case.mjs`? Don't we still need to account for these as regression test cases?

mohd-akram replied:
The test was wrong, it was using exports that are not supposed to be there. We also added other tests that check for more things and added testing `got` directly, so the `got-alike` test can be removed altogether but it doesn't hurt to have. There's also already another `default-class.mjs`.

timfish replied:
This could be kept but it doesn't serve much of a purpose any more.

`snake_case.mjs` only exports a default export and was imported elsewhere with `export * from` which means it's dropped and not exported and there are multiple other places this is tested. It existed before to ensure namespaces had valid identifier names which were derived from file names.

AbhiPrasad reviewed (approved):
Okay seems reasonable enough to me. I tested with some fixtures in https://github.com/nodejs/node/tree/main/test/fixtures/es-modules and tried applying the patch on some sentry internal projects and nothing crashed, let's give it a try!
